### PR TITLE
Draft of arduino reading A1 voltage when commanded by I2C

### PR DESCRIPTION
**Pull request by Ryanf55**
This code is intended to provide the voltage of pin A1 for voltage sensing for support of #61 .When a specific node has an io command of READ_VOLTAGE, the correct node should reply with the value from an analog read of A1. 

I do not have the hardware to test this, but it at least compiles. Please test it. 

It also requires some python script to accomplish the following:

1. Configure which node has voltage sensing (if any)

1. Poll that node for the analog voltage readout

1. Use a voltage divider ratio from something like the program options to compute the battery voltage.

Later, alerts could be added and cell count could be autodetected to alert if the lap timer is low on battery.

### CODE REVIEW

File: src/node/node.ino [hunk 1]
```diff
@@ -815,6 +823,10 @@ void i2cTransmit()
             ioBufferWrite32(millis());
             break;
 
+        case READ_VOLTAGE: // RF: READ_VOLTAGE will get the current node voltage from the defined analog pin
+            ioBufferWrite32(voltageRead());
```
Comment (Ryanf55) [inline src/node/node.ino]: Not sure which ioBufferWrite function is needed.

File: src/node/node.ino [hunk 2]
```diff
@@ -377,6 +379,12 @@ int rssiRead()
     return analogRead(0);
 }
 
+//RF: Read the voltage of the node battery voltage divider on A1
+int voltageRead()
+{
+  return analogRead(1);
```
Comment (Ryanf55) [inline src/node/node.ino]: I hard coded A1. I don't believe this is being used for anything

File: src/node/node.ino (hunk 1 above)
Comment (HazardCreative) [reply src/node/node.ino]: Analog reads are 10-bit, so the raw data would fit best in the 16-bit write.

```suggestion
            ioBufferWrite16(voltageRead()); 
```

**Comment by HazardCreative** [comment]
I like the idea of doing this without much additional hardware.
* To make things easy on the user, we should specify the resistor sizes needed on the divider (and preferably match as many to what they are already buying as possible) and provide good defaults on the divider.
* It would be best to only turn the divider on when a reading is asked for so that isn't just sinking current.
* On the interface side, this should probably live in the same routines as the temp sensor work that was merged in recently. That will help when we write a 'health check' that can alert the user of values for any of these sensors.
* Perhaps which node has voltage sensing could be automatic, figured out by checking the values on each during server initialization. Users could potentially configure as many different sensors as they have nodes if desired.

**Comment by Ryanf55** [comment]
Ok. 

- For a 5V or 3.3V logic Arduino and up to 6s voltage input, using resistors from the [Sparkfun resistor kit](https://www.sparkfun.com/products/10969), a 10kΩ and 100kΩ would do the trick. (2.29V output at full 6s battery). I'd like to use just the 1k and 100k, but it takes at least four of them to achieve a decent output because the ratio is so large. 

- That's not an issue, through the 10k-100k divider, the current flow at 6s battery voltage is 0.23mA. 

- Sounds good with me. A programmable voltage minimum would be good, or at least lipo cell count selector. Between 5s and 6s, there is too much overlap to be able to guess what kind of battery it is unless previous data is used

- Wouldn't all the nodes be powered off the same battery? I'm not sure how to detect automatic. I can look at the analogRead to see if it is floating or pulled somewhere. If it's floating, I do not think auto sensing would work reliably. However, it may not be. I can test this and report back.

**Comment by HazardCreative** [comment]
Thanks for checking on using the parts a user might already be buying a bunch of. Your proposed values would only give us a resolution of ~0.05V, but that's probably good enough for the purposes we're using it for.

Could you enable the internal pull-up to cancel the float? If you're worried about it affecting the reading, turn it on briefly during the check and then disable it after?

**Comment by Ryanf55** [comment]
Right. I am happy to write some additional documentation sharing alternate resistor values depending on the hardware and max voltage Li-Po used if resolution is more important than BOM cost/complexities to cover the following cases. Either situation will use resistors from that Sparkfun kit as it has the most common resistors. 

1. 5V logic, up to 6s input
1. 5V logic, up to 3s input
1. 3.3V logic, up to 6s input
1. 3.3V logic, up to 3s input


After some research, it appears internal pull-up works OK, however we will need to use lower resistor values so it doesn't affect the readings as much. Even with lower resistance values, the current draw is still minuscule. 

I just tried it on my Uno, and it floats from 1019 - 1020 with no resistors connected. Try it with a nano if you can to confirm my results. If the results are true, the input pull-up will be a valid approach to check if a node has voltage sensing hooked up. During initialization, a request of that value will be done. If it's below something like 1015, then the software can add that node to a list of nodes that have voltage sensing.

```c
int pin_voltage_sense = A1;

void setup() {
  pinMode(pin_voltage_sense,INPUT_PULLUP);
  Serial.begin(9600);
  
}

void loop() {
  Serial.println(analogRead(pin_voltage_sense));

}
```

**Comment by HazardCreative** [comment]
I couldn't quickly find good information on how the internal pullup affected the values, just some "it depends" answers without any information on how to calculate it. Maybe the best approach is to disable it after the initial check, before you start taking readings.

I believe pinMode(pin_voltage_sense,INPUT) explicitly disables it, so this shouldn't be too hard.

**Comment by Ryanf55** [comment]
Same. 

Here's some code to check in setup what the value is, then set to normal pullup. 

```
int pin_voltage_sense = A1;
bool equipped_with_vsense = false;

void setup() {
  //Check if a node is connected by setting the internal pullup
  pinMode(pin_voltage_sense,INPUT_PULLUP);
  
  if (analogRead(pin_voltage_sense) < 1015) {
    //voltage sensing must be connected
    equipped_with_vsense = true;
    //turn off internal pullup for more accurate readings
    pinMode(pin_voltage_sense,INPUT);
  }

  
  Serial.begin(9600);
  
}

void loop() {
  Serial.println(analogRead(pin_voltage_sense));

}
```

Once I get the hardware, I'll check it actually works like I think with a precise power supply.

**Comment by HazardCreative** [comment]
I appreciate the work on this, but I think for the moment I'm going to leave voltage sensing to an external sensor. This keeps the load down on the nodes so they can be dedicated to sensing, which may become even more important if progress toward #72 begins. #89 provides a good solution for the problem this is trying to solve.

**Pull request closed without merging**